# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/IonBuilder.cpp
# Commit: 307481f32d44
# Full Hash: 307481f32d4403dae943b9ff26cd7a7e2eadf083
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2019-12-15 21:49:48
# Regressor Bug: 1601599
# File Overlap Count: 1
# Description:
#   Bug 1601599 part 1 - Fix assert in IonBuilder::setStaticName to account for loop phis. r=tcampbell
#   
#   This is necessary once we support OSR for spread expressions.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D56698
# ==============================================================================

diff -r 273abf200022 -r 307481f32d44 js/src/jit/IonBuilder.cpp
--- a/js/src/jit/IonBuilder.cpp	Sun Dec 15 10:22:50 2019 +0000
+++ b/js/src/jit/IonBuilder.cpp	Sun Dec 15 11:32:25 2019 +0000
@@ -8568,9 +8568,12 @@
 
   current->pop();
 
-  // Pop the bound object on the stack.
+  // Pop the bound object on the stack. This is usually a constant but it can
+  // be a phi if loops are involved, for example: x = [...arr];
   MDefinition* obj = current->pop();
-  MOZ_ASSERT(&obj->toConstant()->toObject() == staticObject);
+  MOZ_ASSERT(obj->isConstant() || obj->isPhi());
+  MOZ_ASSERT_IF(obj->isConstant(),
+                &obj->toConstant()->toObject() == staticObject);
 
   if (needsPostBarrier(value)) {
     current->add(MPostWriteBarrier::New(alloc(), obj, value));
